library/kani/src/shadow.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+extern "C" {
+    pub fn __KANI_pointer_object(ptr: *const u8) -> usize;
+    pub fn __KANI_pointer_offset(ptr: *const u8) -> usize;
+}
+
+const MAX_NUM_OBJECTS: usize = 1024;
+const MAX_OBJECT_SIZE: usize = 64;
+
+pub struct ShadowMem {
+    is_init: [[bool; MAX_OBJECT_SIZE]; MAX_NUM_OBJECTS],
+}
+
+impl ShadowMem {
+    pub const fn new() -> Self {
+        Self { is_init: [[false; MAX_OBJECT_SIZE]; MAX_NUM_OBJECTS] }
+    }
+
+    /// # Safety
+    ///
+    /// `ptr` must be valid
```

Comment:
Why raw pointer? Why does it need to be valid?

Reply:
> Why raw pointer?

Do you suggest using a reference instead?

> Why does it need to be valid?

I thought `__CPROVER_POINTER_OBJECT` required a valid pointer, but the [CBMC docs](https://diffblue.github.io/cbmc/cprover-manual/md_memory_primitives.html) indicates it doesn't have any preconditions.

Reply:
I think the raw pointer interface makes sense if you want to be able to reason about potentially invalid pointer. The initialization case is definitely one where you do want a raw pointer exactly for that reason.

Reply:
That said, I don't think this function is unsafe or require valid pointer. So making this clear is probably the best outcome

Reply:
Removed `unsafe` and validity requirement.